// @aws-sdk/client-apigatewaymanagementapi/dist-es/models/models_0.js
import { __assign } from "tslib";
export var DeleteConnectionRequest;
(function (DeleteConnectionRequest) {
    DeleteConnectionRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteConnectionRequest || (DeleteConnectionRequest = {}));
export var ForbiddenException;
(function (ForbiddenException) {
    ForbiddenException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ForbiddenException || (ForbiddenException = {}));
export var GoneException;
(function (GoneException) {
    GoneException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GoneException || (GoneException = {}));
export var LimitExceededException;
(function (LimitExceededException) {
    LimitExceededException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(LimitExceededException || (LimitExceededException = {}));
export var GetConnectionRequest;
(function (GetConnectionRequest) {
    GetConnectionRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetConnectionRequest || (GetConnectionRequest = {}));
export var Identity;
(function (Identity) {
    Identity.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(Identity || (Identity = {}));
export var GetConnectionResponse;
(function (GetConnectionResponse) {
    GetConnectionResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetConnectionResponse || (GetConnectionResponse = {}));
export var PayloadTooLargeException;
(function (PayloadTooLargeException) {
    PayloadTooLargeException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PayloadTooLargeException || (PayloadTooLargeException = {}));
export var PostToConnectionRequest;
(function (PostToConnectionRequest) {
    PostToConnectionRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PostToConnectionRequest || (PostToConnectionRequest = {}));
