// @aws-sdk/client-apigatewaymanagementapi/dist-cjs/models/models_0.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PostToConnectionRequest = exports.PayloadTooLargeException = exports.GetConnectionResponse = exports.Identity = exports.GetConnectionRequest = exports.LimitExceededException = exports.GoneException = exports.ForbiddenException = exports.DeleteConnectionRequest = void 0;
var DeleteConnectionRequest;
(function (DeleteConnectionRequest) {
    DeleteConnectionRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(DeleteConnectionRequest = exports.DeleteConnectionRequest || (exports.DeleteConnectionRequest = {}));
var ForbiddenException;
(function (ForbiddenException) {
    ForbiddenException.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(ForbiddenException = exports.ForbiddenException || (exports.ForbiddenException = {}));
var GoneException;
(function (GoneException) {
    GoneException.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GoneException = exports.GoneException || (exports.GoneException = {}));
var LimitExceededException;
(function (LimitExceededException) {
    LimitExceededException.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(LimitExceededException = exports.LimitExceededException || (exports.LimitExceededException = {}));
var GetConnectionRequest;
(function (GetConnectionRequest) {
    GetConnectionRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetConnectionRequest = exports.GetConnectionRequest || (exports.GetConnectionRequest = {}));
var Identity;
(function (Identity) {
    Identity.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(Identity = exports.Identity || (exports.Identity = {}));
var GetConnectionResponse;
(function (GetConnectionResponse) {
    GetConnectionResponse.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(GetConnectionResponse = exports.GetConnectionResponse || (exports.GetConnectionResponse = {}));
var PayloadTooLargeException;
(function (PayloadTooLargeException) {
    PayloadTooLargeException.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(PayloadTooLargeException = exports.PayloadTooLargeException || (exports.PayloadTooLargeException = {}));
var PostToConnectionRequest;
(function (PostToConnectionRequest) {
    PostToConnectionRequest.filterSensitiveLog = (obj) => ({
        ...obj,
    });
})(PostToConnectionRequest = exports.PostToConnectionRequest || (exports.PostToConnectionRequest = {}));
